test/functional/lib/tests.cpp

```diff
@@ -6567,6 +6771,83 @@ GenericTxChecksumOffloadTcp(
     MpTxFlush(GenericMp);
 }
 
+VOID
+GenericRxChecksumOffloadTcp(
+    ADDRESS_FAMILY Af
+) {
+    const BOOLEAN Rx = TRUE, Tx = FALSE;
+    auto If = FnMp1QIf;
+    UINT16 LocalPort, RemotePort;
+    ETHERNET_ADDRESS LocalHw, RemoteHw;
+    INET_ADDR LocalIp, RemoteIp;
+    UINT32 AckNum = 0;
+    auto Xsk = CreateAndBindSocket(If.GetIfIndex(), If.GetQueueId(), Rx, Tx, XDP_GENERIC);
+    RemotePort = htons(4321);
+    auto TcpSocket = CreateTcpSocket(Af, &If, &LocalPort, RemotePort, &AckNum);
+    auto GenericMp = MpOpenGeneric(If.GetIfIndex());
+
+    // Routine Description:
+    //     This test verifies the metadata that indicates RX checksum offloads were done for TCP.
+    //     It does so by injecting a TCP frame with a valid TCP checksum and verifying that
+    //     the metadata is correctly set by the miniport driver.
+
+    EnableRxChecksumOffload(&Xsk);
+    ActivateSocket(&Xsk, Rx, Tx);
+
+    Xsk.RxProgram =
+        SocketAttachRxProgram(If.GetIfIndex(), &XdpInspectRxL2, If.GetQueueId(), XDP_GENERIC, Xsk.Handle.get());
+
+    If.GetHwAddress(&LocalHw);
+    If.GetRemoteHwAddress(&RemoteHw);
+    if (Af == AF_INET) {
+        If.GetIpv4Address(&LocalIp.Ipv4);
+        If.GetRemoteIpv4Address(&RemoteIp.Ipv4);
+    } else {
+        If.GetIpv6Address(&LocalIp.Ipv6);
+        If.GetRemoteIpv6Address(&RemoteIp.Ipv6);
+    }
+
+    // Construct a valid IPv4/IPv6 + TCP frame with a valid TCP checksum
+    UCHAR TcpPayload[] = "GenericRxChecksumOffloadTcp";
+    UCHAR TcpFrame[TCP_HEADER_STORAGE + sizeof(TcpPayload)];
+    UINT32 TcpFrameLength = sizeof(TcpFrame);
+
+    TEST_TRUE(
+        PktBuildTcpFrame(
+            TcpFrame, &TcpFrameLength, TcpPayload, sizeof(TcpPayload), NULL, 0, 0, 0, TH_SYN, 0,
+            &LocalHw, &RemoteHw, Af, &LocalIp, &RemoteIp, LocalPort, RemotePort));
+
+    RX_FRAME RxFrame;
+    RxInitializeFrame(&RxFrame, If.GetQueueId(), TcpFrame, TcpFrameLength);
+    RxFrameSetChecksumOffloadState(&RxFrame, TRUE, FALSE, FALSE, FALSE, FALSE, FALSE);
```

Comment:
Maybe we should create a subroutine that takes all these FNMP offload parameters along with the expected XDP_FRAME_CHECKSUM and XDP_FRAME_LAYOUT values to verify against, and the subroutine handles all the common code like socket creation and packet indicate/recv?

Reply:
Can we do this in a future refactor PR? Adding this subroutine, I would imagine would simplify many other places in the test code, not just the RX offload side.

Reply:
My own idea was a subroutine specifically for this class of RX checksum offload tests, but I'm open to other refactorings too. Personally, I'd lean towards a subroutine that is more or less a copy of any of these test cases, and then just make the offload settings to FNMP into a function parameter, and same for the XDP_FRAME_CHECKSUM.